### PR TITLE
Dumbest model

### PR DESCRIPTION
**Pull request by laurieskelly**
eats a kafka message, extracts the budget questions, gets the "midpoint" of the budget answer ranges, and averages them with no weighting. 

predicting "per item spend" in a completely coarse and naive way.

### CODE REVIEW

File: src/models/dumbest_model.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+
+# local
+from ..data.kafka_ingestor import KafkaIngestor
+from ..features.budget_midpoints import budget_midpoints
+
```
Comment (jscottcronin) [inline src/models/dumbest_model.py]: import from root directory

File: src/models/dumbest_model.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+class ReallyDumbModel(object):
+    '''Predicts per-item spend by averaging the midpoints of the onboarding budget
+    questions.
+    '''
+
+    def budget_from_kafkafile(self, kafkafile):
+        ki = KafkaIngestor(kafkafile)
+        return budget_midpoints(ki.data)
+
```
Comment (jscottcronin) [inline src/models/dumbest_model.py]: should this be a private method?

File: src/models/dumbest_model.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+    def weights_from_features(self, features):
+        '''returns an nparray the same length as features with identical elements
```
Comment (jscottcronin) [inline src/models/dumbest_model.py]: private method?

File: src/models/dumbest_model.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+        that sum to one.
+        ex: features = [1, 2, 3] --> [0.33, 0.33, 0.33]
+        '''
+        return np.ones_like(features)/float(len(features))
+
+
+    def predict(self, kafkafile):
+        '''returns the average of the "midpoints" of the answers to the budget
```
Comment (jscottcronin) [inline src/models/dumbest_model.py]: lets put the public method at top of class to make more readable

File: src/models/dumbest_model.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        questions with no weighting
+        '''
```
Comment (jscottcronin) [inline src/models/dumbest_model.py]: can we add arguments and return for public method doc string

File: src/models/dumbest_model.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        feature_dict = self.budget_from_kafkafile(kafkafile)
+        features = np.array(feature_dict.values())
+
+        weights = self.weights_from_features(features)
+        return np.dot(features, weights)
```
Comment (jscottcronin) [inline src/models/dumbest_model.py]: 💯 🙏

File: src/data/kafka_ingestor.py [hunk 7]
```diff
@@ -28,7 +30,8 @@ def _clean_profile_answers(self):
             color = 'dislike_color' in question
             age = 'age_range' in question
             budget = 'budget' in question
-            if color or age or budget:
+            stores = 'stores' in question
+            if color or age or budget or stores:
```
Comment (TCJeff) [inline src/data/kafka_ingestor.py]: this whole block reads kinda weird to me, your looking for the presence of keys in a dictionary and then adding the answers if they were in there?
ruby has something like this:
```
valid_question_keys = ['dislike_color', 'age_range', 'budget', 'stores']

raw_profile.each_with_object({}) do |item, profile_dict|
  question = item['question_id']
  next unless valid_question_keys.include?(question)
  profile_dict[question] = item['answers']
end
```
does python have anything similar? then you only have to worry about writing one word instead 2 lines every time you want to make a change going forward!

File: src/features/budget_midpoints.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+def budget_midpoints(onbo_data):
+    '''takes budget columns from onboarding data and returns a dictionary, keys
+    being the budget items and values the midpoints of the ranges of the answers.
+    '''
+
+    midpoints = {}
+    for key, value in onbo_data.iteritems():
+        if 'budget' in key:
+            budget_label = clean_question_label(key)
+            budget_value = get_range_midpoint(value)
```
Comment (TCJeff) [inline src/features/budget_midpoints.py]: you should probably sanitize/normalize the budget values too since they are technically humanly input and can change in the future

File: src/models/dumbest_model.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+import numpy as np
+
+# local
+from ..data.kafka_ingestor import KafkaIngestor
+from ..features.budget_midpoints import budget_midpoints
+
+class ReallyDumbModel(object):
```
Comment (TCJeff) [inline src/models/dumbest_model.py]: pls consider a different name.... or at least put in what you're trying to model

File: src/data/kafka_ingestor.py (hunk 7 above)
Comment (laurieskelly) [inline src/data/kafka_ingestor.py]: haha, this started as something that was maybe clever and then went all frankenstein. the reason I did it this weird way was that I wanted each of many possible keys that are strings with the substring "budget" in them. 

when I realized i also wanted other things, I should have added those other keys in a more rational way. at this point i was just playing around with doing complicated booleans in this stepwise way. good catch

File: src/data/kafka_ingestor.py [hunk 10]
```diff
@@ -1,4 +1,6 @@
-from settings import *
+import json
+
+from ..settings import *
```
Comment (laurieskelly) [inline src/data/kafka_ingestor.py]: these are all relative imports within the same package so this is fine. as corroboration, here's a random scikit-learn module, it's the first one i looked at and it has several in its imports:
https://github.com/scikit-learn/scikit-learn/blob/14031f6/sklearn/feature_selection/univariate_selection.py#L422

& I would be more explicit with outside packages but it's fine to import * from my own settings, there's no danger of namespace collision here

File: src/features/budget_midpoints.py (hunk 8 above)
Comment (laurieskelly) [inline src/features/budget_midpoints.py]: good idea

File: src/models/dumbest_model.py (hunk 1 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: same as above

File: src/models/dumbest_model.py (hunk 2 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: added underscore

File: src/models/dumbest_model.py (hunk 3 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: added underscore

File: src/models/dumbest_model.py (hunk 4 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: moved to top

File: src/models/dumbest_model.py (hunk 9 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: i will consider it. but i make no promises. it says what it does in the comment on line 8 tho

File: src/models/dumbest_model.py (hunk 5 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: done

File: src/models/dumbest_model.py (hunk 6 above)
Comment (laurieskelly) [inline src/models/dumbest_model.py]: i assume this is good! anything in particular that you like?

Comment (jscottcronin) [inline src/models/dumbest_model.py]: simple and clear. very easy to follow. nice job :)

File: src/features/budget_midpoints.py (hunk 8 above)
Comment (laurieskelly) [inline src/features/budget_midpoints.py]: @TCJeff can you look at what i did in the most recent commit? 
I added a couple of validations with assert statements. It feels sort of half-assed and shoved in there but I am not sure how I might do a better job

Comment (laurieskelly) [inline src/features/budget_midpoints.py]: also, meta question @TCJeff -- if there's a better way to conveniently point to the place in the code where I fixeded it, I am all ears

Comment (TCJeff) [inline src/features/budget_midpoints.py]: https://github.com/trunkclub/ds_price_sensitivity_workspace/blob/c123a30988f7cce7c02e2f68c78e8d88d36e573f/src/features/budget_midpoints.py#L44-L47

click view -> shift click line numbers on the left and you get the anchor

File: src/features/budget_midpoints.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+        return int(a)/2 + int(a)
+
+    else:
+        msg = "Could not parse answer format. " \
+              "Expecting format like ['50-100'] or ['200+']"
+        raise ValueError(msg)
+
+
+def validate_budget_answer(answer):
+    assert type(answer) is list, "Answer format is %s, expected list." % type(answer)
```
Comment (TCJeff) [inline src/features/budget_midpoints.py]: yeah this does feel weird... in ruby i would do something like
```
def validate_budget_answer(answer)
  ranges = answer.split('-')
  return ranges.map{|i| Integer(i)} if ranges.length == 2
  max_val = answers.split('+')
  raise UnexpectedInputError if max_val.length != 1  # this is a little weird since it will also raise an error if whatever value present isn't a valid integer
  val = Integer(max_val)
  return [val, val*2]
end
```

now the output interface is always the same and will give you consistent values and you can also skip the other split in your get range midpoint!
